### PR TITLE
[Xcode] Compute PGO profdata paths instead of searching for them at build time

### PR DESCRIPTION
**Pull request by emw-apple**
#### 6dfb5dc2b3bcd083231cee3cacc599ca7f76998b
<pre>
[Xcode] Compute PGO profdata paths instead of searching for them at build time
<a href="https://bugs.webkit.org/show_bug.cgi?id=239681">https://bugs.webkit.org/show_bug.cgi?id=239681</a>

Reviewed by Wenson Hsieh and Alexey Proskuryakov.

Production builds should always use profdata from the SDK, internal
release builds should use profdata from the local WebKitAdditions when
available, and open source builds should use the stub profdata. Add a
PROFILE_DATA_PATH build setting which implements these rules and replace
the script phases that checked those paths dynamically.

This fixes XCBuild validation issues in the open-source build, and
should prevent production builders from using stale profdata in
BUILT_PRODUCTS_DIR. Additionally, engineers can now build with custom
profile data by overriding PROFILE_DATA_PATH.

* Source/JavaScriptCore/Configurations/JavaScriptCore.xcconfig:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Scripts/copy-profiling-data.sh: Added. This file
  is part of JavaScriptCore&apos;s private headers and replaces the &quot;Copy
  Profiling Data&quot; script phase logic all projects.
* Source/WebKit/Configurations/BaseTarget.xcconfig:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/251060@main">https://commits.webkit.org/251060@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@294952">https://svn.webkit.org/repository/webkit/trunk@294952</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by emw-apple** [comment]
This is a reland of #367. It fixes a scripting bug and preserves the internal-engineering-without-PGO build workflow.

File: Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -12106,9 +12110,9 @@
...
 			);
 			inputPaths = (
-				"$(BUILT_PRODUCTS_DIR)/usr/local/include/WebKitAdditions/Profiling/JavaScriptCore.profdata.compressed",
+				"$(PROFILE_DATA_PATH)",
 				"$(SRCROOT)/../../Tools/Profiling/Empty.profdata",
-				"$(SDK_DIR)/usr/local/include/WebKitAdditions/Profiling/JavaScriptCore.profdata.compressed",
+				"$(SRCROOT)/Scripts/copy-profiling-data.sh",
 			);
 			name = "Copy Profiling Data";
 			outputFileListPaths = (
```
Comment (aproskuryakov) [inline Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:12118]: Could you please double check that there are output files? Not seeing them on GitHub?

File: Source/JavaScriptCore/Scripts/copy-profiling-data.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/sh
+
+if [ -z "${PROFILE_DATA_FLAGS}" ]; then
+    exit 0;
+fi
+
+src="${SCRIPT_INPUT_FILE_0}"
```
Comment (aproskuryakov) [inline Source/JavaScriptCore/Scripts/copy-profiling-data.sh]: I feel like longer and more descriptive names would be more in line with WebKit coding style.

**Review by aproskuryakov — APPROVED** [review]

File: Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj (hunk 1 above)
Comment (emw-apple) [reply Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:12118]: Yes, I think GitHub is struggling to expand the common lines of the patch. Here is the whole script phase object:

		F4CDF3C927E9147500191928 /* Copy Profiling Data */ = {
			isa = PBXShellScriptBuildPhase;
			buildActionMask = 12;
			files = (
			);
			inputFileListPaths = (
			);
			inputPaths = (
				"$(PROFILE_DATA_PATH)",
				"$(SRCROOT)/../../Tools/Profiling/Empty.profdata",
				"$(SRCROOT)/Scripts/copy-profiling-data.sh",
			);
			name = "Copy Profiling Data";
			outputFileListPaths = (
			);
			outputPaths = (
				"$(BUILT_PRODUCTS_DIR)/DerivedSources/JavaScriptCore/JavaScriptCore.profdata",
			);
			runOnlyForDeploymentPostprocessing = 0;
			shellPath = /bin/sh;
			shellScript = "if [ \"${ACTION}\" = \"installhdrs\" -o \"${ACTION}\" = \"installapi\" ]; then\n    exit 0;\nfi\n${SCRIPT_INPUT_FILE_2}\n";
		};

File: Source/WebKit/Configurations/BaseTarget.xcconfig [hunk 3]
```diff
@@ -50,6 +50,12 @@ ALTERNATE_HEADER_SEARCH_PATHS_iosmac = $(BUILT_PRODUCTS_DIR)$(WK_ALTERNATE_FRAME
 
 HEADER_SEARCH_PATHS = $(ALTERNATE_HEADER_SEARCH_PATHS) $(BUILT_PRODUCTS_DIR)/usr/local/include "$(WEBCORE_PRIVATE_HEADERS_DIR)/ForwardingHeaders" $(BUILT_PRODUCTS_DIR)/DerivedSources/WebKit $(WEBGPU_HEADER_SEARCH_PATHS) $(WEBKITADDITIONS_HEADER_SEARCH_PATHS) $(LIBWEBRTC_HEADER_SEARCH_PATHS) $(SRCROOT) $(HEADER_SEARCH_PATHS);
 
+PROFILE_DATA_PATH = $(PROFILE_DATA_PATH_PRODUCTION_$(DEPLOYMENT_LOCATION));
```
Comment (whsieh) [inline Source/WebKit/Configurations/BaseTarget.xcconfig]: Just to double check — `$(DEPLOYMENT_LOCATION)` here is set to `YES`/`NO` based on whether or not it's a production build?

**Review by whsieh — APPROVED** [review]
Looks great to me!

File: Source/WebKit/Configurations/BaseTarget.xcconfig (hunk 3 above)
Comment (aproskuryakov) [reply Source/WebKit/Configurations/BaseTarget.xcconfig]: Hmm, this looks like it might be wrong for builds that install to device.

Comment (emw-apple) [reply Source/WebKit/Configurations/BaseTarget.xcconfig]: I think Alexey is right: DEPLOYMENT_LOCATION should be YES for _any_ install build, including builds that install to device.

I wonder if the correct thing here would be to check the configuration, i.e. 

    DEPLOYMENT_LOCATION[config=Production] = $(SDK_DIR)...

Any objections to that?

Comment (whsieh) [reply Source/WebKit/Configurations/BaseTarget.xcconfig]: I see! Yeah, my followup question was going to be why we preferred that over using `$(CONFIGURATION)` or `[config=Production]`. It sounds like we'll want to switch to one of the two?

Comment (aproskuryakov) [reply Source/WebKit/Configurations/BaseTarget.xcconfig]: No objections from me. I wonder if we have a way to test install to device builds in CI in the future.

Comment (emw-apple) [reply Source/WebKit/Configurations/BaseTarget.xcconfig]: The new stack is:

    PROFILE_DATA_PATH = $(PROFILE_DATA_PATH_INTERNAL_$(USE_INTERNAL_SDK));
    PROFILE_DATA_PATH_INTERNAL_ = $(SRCROOT)/../../Tools/Profiling/Empty.profdata;
    PROFILE_DATA_PATH_INTERNAL_YES = $(BUILT_PRODUCTS_DIR)/usr/local/include/WebKitAdditions/Profiling/JavaScriptCore.profdata.compressed;
    PROFILE_DATA_PATH_INTERNAL_YES[config=Production] = $(SDK_DIR)/usr/local/include/WebKitAdditions/Profiling/JavaScriptCore.profdata.compressed;

Thanks for catching this!

**Pull request merged**